flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamIterationHead.java

```diff
@@ -17,100 +17,164 @@
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * TODO write javadoc
+ * <p>
+ * - open a list state per snapshot process
+ * - book-keep snapshot logs
+ * - Clean up state when a savepoint is complete - ONLY in-transit records who do NOT belong in other snapshots
+ *
+ * @param <IN>
+ */
 @Internal
-public class StreamIterationHead<OUT> extends OneInputStreamTask<OUT, OUT> {
+public class StreamIterationHead<IN> extends OneInputStreamTask<IN, IN> {
 
 	private static final Logger LOG = LoggerFactory.getLogger(StreamIterationHead.class);
 
 	private volatile boolean running = true;
 
-	// ------------------------------------------------------------------------
-	
+	private volatile RecordWriterOutput<IN>[] outputs;
+
+	private UpstreamLogger<IN> upstreamLogger;
+
+	private Object lock;
+
+	@Override
+	public void init() throws Exception {
+		this.lock = getCheckpointLock();
+		getConfiguration().setStreamOperator(new UpstreamLogger(getConfiguration()));
+		operatorChain = new OperatorChain<>(this);
+		this.upstreamLogger = (UpstreamLogger<IN>) operatorChain.getHeadOperator();
+	}
+
 	@Override
 	protected void run() throws Exception {
-		
+
 		final String iterationId = getConfiguration().getIterationId();
 		if (iterationId == null || iterationId.length() == 0) {
 			throw new Exception("Missing iteration ID in the task configuration");
 		}
-		
-		final String brokerID = createBrokerIdString(getEnvironment().getJobID(), iterationId ,
-				getEnvironment().getTaskInfo().getIndexOfThisSubtask());
-		
+		final String brokerID = createBrokerIdString(getEnvironment().getJobID(), iterationId,
+			getEnvironment().getTaskInfo().getIndexOfThisSubtask());
 		final long iterationWaitTime = getConfiguration().getIterationWaitTime();
 		final boolean shouldWait = iterationWaitTime > 0;
 
-		final BlockingQueue<StreamRecord<OUT>> dataChannel = new ArrayBlockingQueue<StreamRecord<OUT>>(1);
+		final BlockingQueue<Either<StreamRecord<IN>, CheckpointBarrier>> dataChannel
+			= new ArrayBlockingQueue<>(1);
 
 		// offer the queue for the tail
 		BlockingQueueBroker.INSTANCE.handIn(brokerID, dataChannel);
 		LOG.info("Iteration head {} added feedback queue under {}", getName(), brokerID);
 
 		// do the work 
 		try {
-			@SuppressWarnings("unchecked")
-			RecordWriterOutput<OUT>[] outputs = (RecordWriterOutput<OUT>[]) getStreamOutputs();
+			outputs = (RecordWriterOutput<IN>[]) getStreamOutputs();
 
 			// If timestamps are enabled we make sure to remove cyclic watermark dependencies
 			if (isSerializingTimestamps()) {
-				for (RecordWriterOutput<OUT> output : outputs) {
+				for (RecordWriterOutput<IN> output : outputs) {
 					output.emitWatermark(new Watermark(Long.MAX_VALUE));
 				}
 			}
 
+			synchronized (lock) {
+				//emit in-flight events in the upstream log upon recovery
+				for (StreamRecord<IN> rec : upstreamLogger.getReplayLog()) {
```

Comment:
Maybe it would make sense to put this in a while loop and check "running" as well, to cancel early if the job is cancelled during replay